collectionService.js

```diff
@@ -410,7 +414,32 @@ var CollectionService = module.exports = BaseService.extend( {
 			this.length++;
 		}
 
-		_.extend( this._recordsById[ recordId ], dto );
+		for( field in dto ) {
+			if( _.isObject( dto[ field ] ) ){
+				this._recordsById[ recordId ][ field ] = _.clone( dto[ field ] );
+				this._deepFreeze( this._recordsById[ recordId ][ field ] );
+			} else this._recordsById[ recordId ][ field ] = dto[ field ];			
+		}
+	},
+
+	// From : https://github.com/jsdf/deep-freeze
+	_deepFreeze( obj ) {
+		Object.freeze( obj );
+	
+		var objIsFunction = typeof obj === "function";
+		var hasOwnProp = Object.prototype.hasOwnProperty;
+	
+		Object.getOwnPropertyNames( obj ).forEach( function ( property ) {
+			if ( hasOwnProp.call( obj, property )
+			&& ( objIsFunction ? property !== 'caller' && property !== 'callee' && property !== 'arguments' : true )
+			&& obj[ prop ] !== null
```

Comment:
this check does not seem necessary since we are already checking `typeof obj[ property ] === "object"`

Reply:
Hi @dgbeck! This is a weird one.  [the data type of null is an object.](https://www.w3schools.com/js/js_typeof.asp#:~:text=In%20JavaScript%20null%20is%20%22nothing,typeof%20null%20is%20an%20object.) So, I left that one there.